pep-9999.rst

```diff
@@ -0,0 +1,464 @@
+efforts), data races on shared mutable objects become a more
+pressing concern. A deep immutability mechanism ensures that
+shared objects are not modified concurrently, enabling safer
+multi-threaded and parallel computation.
+
+See also the discussion about extensions further down in this
+document.
+
+
+Optimisations and Caching Benefits
+----------------------------------
+
+Immutable objects provide opportunities for optimisation, such as
+structural sharing, memoization, and just-in-time (JIT)
+compilation techniques (specialising for immutable data, e.g.
+fixed shape, fewer barriers, inlining, etc.). Freezing objects can
+allow Python to implement more efficient caching mechanisms and
+enable compiler optimisations that rely on immutability
+assumptions.
+
+
+Note that none of the above are *guaranteed* with immortal
+objects as these may still change if a program manipulates an
+object without knowing that it is immortal.
+
+
+Specification
+=============
+
+Changes to Python objects
+-------------------------
+
+Every Python object will have a flag that keeps track of its
+immutability status. Details about the default value of
+this flag is discussed further down in this document. 
+
+The flag can be added without extending the size of the
+Python object header. We initially intend for this flag 
+to be implemented by bit-stealing from the reference 
+counter, but this will change in our planned extensions.
+
+
+Implementation of Immutability
+------------------------------
+
+Immutability is enforced through run-time checking. The macro
+``Py_CHECKWRITE(op)`` is inserted on all paths that are guaranteed
+to end up in a write to ``op``. The macro inspects the immutability
+flag in the header of ``op`` and signals an error if the immutability
+flag is set.
+
+A typical use of this check looks like this:
+
+.. code-block:: c
+
+	if (!Py_CHECKWRITE(op)) {        // perform the check
+	    PyErr_WriteToImmutable(op);  // raise the error if the check fails
+	    return NULL;                 // abort the write
+	}  
+	... // code that performs the write
+
+
+As writes are common but lack a common path that most writes to through
+the PEP requires a ``Py_CHECKWRITE`` call, there are several places in
+the CPython code base that are changed as a consequence of this PEP.
+So far we have identified around 70 writes spread across a dozen files.
+
+
+New Obligations on C Extensions
+-------------------------------
+
+Because our implementation builds on information available to the CPython
+cycle detector, types defined through C code will support immutability 
+"out of the box".
+
+C extensions that perform writes to data which can be made immutable
+should add the ``Py_CHECKWRITE`` macro shown above on all paths in the
+code that lead to writes to that data.
+
+**TODO**
+  If we don't have a way to tell whether or not an extension 
+  supports immutability, then freeze might end up "leaking"
```

Comment:
I think this is likely an unwelcome but needed side effect. There is no way to freeze hidden state inside a c struct without code changes by the extension author.

Reply:
Right, but we need to agree how to handle this. Should an extension writer have to explicitly opt-in to support deep immutability?

Reply:
I think an extension author would need to add `Py_CHECKWRITE` to their code which checks whether the enclosing object is immutable and acts accordingly. We should provide some sample code to that effect, but that process is entirely opt-in.